argus/backend/service/github_service.py

```diff
@@ -0,0 +1,232 @@
+
+LOGGER = logging.getLogger(__name__)
+
+
+class GithubService:
+    LAST_RAN_KEY = "github_service_last_issue_refresh"
+
+    plugins = AVAILABLE_PLUGINS
+    def __init__(self):
+        auth = Auth.Token(token=self.get_installation_token())
+        self.gh = Github(auth=auth, per_page=1000)
+
+
+    def get_plugin(self, plugin_name: str) -> PluginInfoBase | None:
+        return self.plugins.get(plugin_name)
+
+    def get_installation_token(self):
+        self._refresh_installation_token()
+        return current_app.config.get("GITHUB_ACCESS_TOKEN")
+
+    def _refresh_installation_token(self):
+        # TODO: To be replaced by JWT refreshing logic once we have Github App in place
+        pass
+
+    def refresh_stale_issues(self):
+        try:
+            last_ran = RuntimeStore.get(key=self.LAST_RAN_KEY)
+        except RuntimeStore.DoesNotExist:
+            last_ran = RuntimeStore()
+            last_ran.key = self.LAST_RAN_KEY
+            last_ran.value = datetime(year=2020, month=1, day=1, hour=0, minute=0, tzinfo=UTC)
+            last_ran.save()
+
+        LOGGER.info("Starting Github Issue sync...")
+
+        all_issues: list[GithubIssue] = list(GithubIssue.all())
+        issues_by_identifier = { f"{issue.owner.lower()}/{issue.repo.lower()}#{issue.number}": issue for issue in all_issues }
+        update_count = 0
+
+        unique_repos = { f"{issue.owner}/{issue.repo}" for issue in all_issues }
+        for idx, repo in enumerate(unique_repos):
+            LOGGER.info("[%s/%s] Fetching %s...", idx + 1, len(unique_repos), repo)
+            repo = self.gh.get_repo(repo)
+            issues = repo.get_issues(since=last_ran.value, state="all", direction="desc", sort="created")
+            for issue_idx, issue in enumerate(issues):
+                match = re.match(
+                    r"http(s)?://(www\.)?github\.com/(?P<owner>[\w\d]+)/"
+                    r"(?P<repo>[\w\d\-_]+)/(?P<type>issues|pull)/(?P<issue_number>\d+)(/)?",
+                    issue.html_url,
+                )
+
+                identifier = f"{match.group('owner').lower()}/{match.group('repo').lower()}#{match.group('issue_number')}"
+                issue_to_update = issues_by_identifier.get(identifier)
+                if not issue_to_update:
+                    LOGGER.info("[%s/%s] No issue found for %s...", issue_idx + 1, "?", identifier)
+                    continue
+                LOGGER.info("[%s/%s] Refreshing %s...", issue_idx + 1, "?", identifier)
+                issue_to_update.title = issue.title
+                issue_to_update.state = issue.state
+                issue_to_update.labels = [IssueLabel(id=label.id, name=label.name, color=label.color, description=label.description) for label in issue.labels]
+                issue_to_update.assignees = [IssueAssignee(login=assignee.login, html_url=assignee.html_url) for assignee in issue.assignees]
+                issue_to_update.save()
+                update_count += 1
```

Comment:
actually, this number might not be true - if there won't be changes in title/state/labes/assignees then it won't update anything

Reply:
True, I've used this one to check if I got exactly the amount of issues expected, not necessarily updated, so this number represents "Issues found"